w3ctrace/state.go

```diff
@@ -17,109 +18,116 @@ const maxKVPairs = 32
+	listMembers []string
 
-	itemsToFilter := len(states) - maxKVPairs
-	filteredStates := states[:0]
+	// The value for the Instana-specific list member ("in=...").
+	instanaTraceStateValue string
+}
 
-	i := 0
-	for ; itemsToFilter > 0 && i < len(states); i++ {
+// makeState creates a new State with the given values
+func MakeState(listMembers []string, instanaTraceStateValue string) State {
+	return State{
+		listMembers:            listMembers,
+		instanaTraceStateValue: instanaTraceStateValue,
+	}
+}
 
-		if len(states[i]) > thresholdLen {
-			itemsToFilter--
-			continue
+// ParseState parses the value of `tracestate` header. Empty list items are omitted.
+func ParseState(traceStateValue string) (State, error) {
+	listMembers := filterEmptyItems(strings.Split(traceStateValue, ","))
+
+	// Look for the Instana list member first, before discarding any list members due to length restrictions.
+	instanaTraceStateValue := ""
+	instanaTraceStateIdx := -1
+	for i, vd := range listMembers {
+		matchResult := instanaListMemberRegex.FindStringSubmatch(vd)
+		if len(matchResult) == 2 {
+			instanaTraceStateValue = strings.TrimSpace(matchResult[1])
+			instanaTraceStateIdx = i
+			break
 		}
-
-		filteredStates = append(filteredStates, states[i])
 	}
-	filteredStates = append(filteredStates, states[i:]...)
 
-	if len(filteredStates) > maxKVPairs {
-		return filteredStates[:maxKVPairs], nil
+	if instanaTraceStateIdx >= 0 {
+		// remove the entry for instana from the array of list members
+		listMembers = append(listMembers[:instanaTraceStateIdx], listMembers[instanaTraceStateIdx+1:]...)
 	}
 
-	return filteredStates, nil
-}
-
-// Add returns a new state prepended with provided vendor-specific data. It removes any existing
-// entries for this vendor and returns the same state if vendor is empty. If the number of entries
-// in a state reaches the MaxStateEntries, rest of the items will be truncated
-func (st State) Add(vendor, data string) State {
-	if vendor == "" {
-		return st
+	// Depending on whether we found an Instana list member, we can either allow 31 or 32 list members from other vendors.
+	maxListMembers := maxKVPairs
+	if instanaTraceStateValue != "" {
+		maxListMembers--
 	}
 
-	newSt := make(State, 1, len(st)+1)
-	newSt[0] = vendor + "=" + data
-	newSt = append(newSt, st.Remove(vendor)...)
-
-	// truncate the state if it reached the max number of entries
-	if len(newSt) > MaxStateEntries {
-		newSt = newSt[:MaxStateEntries]
+	if len(listMembers) < maxListMembers {
+		return State{listMembers: listMembers, instanaTraceStateValue: instanaTraceStateValue}, nil
 	}
 
-	return newSt
-}
-
-// Fetch retrieves stored vendor-specific data for given vendor
-func (st State) Fetch(vendor string) (string, bool) {
-	i := st.Index(vendor)
-	if i < 0 {
-		return "", false
+	itemsToFilter := len(listMembers) - maxListMembers
+	filteredStates := listMembers[:0]
```

Comment:
`filteredStates` is an empty slice of strings, which is later replaced by `append`. So
In this case, I think it's safe to create a new one, no? So it should be safe to use `var filteredStates []string` or `filteredStates := []string{}`.

Also, a slice still references to its original array/slice, so a change in a slice of a slice/array will modify the original one as well. Which shouldn't happen here, as `filteredStates` doesn't reference any item in `listMembers`.

Reply:
For the record, this code was exactly in this shape and form before my changes: https://github.com/instana/go-sensor/blob/master/w3ctrace/state.go#L30-L47

Basically the only change here is that calculate maxListMembers beforehand (either as maxKVPairs or maxKVPairs-1) and use that instead of maxKVPairs. Plus, I renamed `states` to `listMembers`.

But I'm fine with improving the code as part of this PR. I am not sure if I understand your suggestion, though.

So, the initial `listMembers[:0]` clears out the existing array. Maybe Andrey chose this over `[]string{}` for performance reasons? Quick googling suggests that `listMembers[:0]` will keep the already allocated memory and our expectation is probably that in most cases, we will end up with the same array as before anyway. (Kicking out a list member for being longer than 128 characters should be rare.)

WDYT?

Reply:
I additionally renamed filteredStates to filteredListMembers now, should have done that in the first place.

Reply:
About this:
> Also, a slice still references to its original array/slice, so a change in a slice of a slice/array will modify the original one as well. Which shouldn't happen here, as filteredStates doesn't reference any item in listMembers.

AFAICT, the listMembers is created locally in the function ParseState and discarded after we are done with it, we only return the filteredListMembers. So my thinking is that it shouldn't make a difference whether we make destructive changes to the original listMembers array via filteredStates (now renamed to filteredListMembers) or not.

Reply:
Good point. I didn't notice that it was like that before. You are correct. whatever is written in the second array will be passed to the old one, overwriting its content from the beginning.
Nevermind what I said. the array[:0] trick must stay there.

Reply:
> I didn't notice that it was like that before.

Yeah, it is quite impossible to see that from the diff :-/